src/solver.rs

```diff
@@ -0,0 +1,134 @@
+        // script.
+        [op::RETURN, rest @ ..] => rest
+            .iter()
+            .map(|op| op.extract_push_value().ok().cloned())
+            // This `.collect()` is equivalent to `script::Opcode::is_push_only` for the
+            // subset of inputs consisting of entirely valid `PushValue`s, which is all we
+            // care about here.
+            .collect::<Option<_>>()
+            .map(|data| ScriptKind::NullData { data }),
+
+        // Legacy Pay-to-Public-Key (P2PK), which only occurred in early Zcash coinbase outputs.
+        [Opcode::PushValue(PushValue::LargeValue(PushdataBytelength(v))), op::CHECKSIG] => {
+            (v.len() == PubKey::SIZE || v.len() == PubKey::COMPRESSED_SIZE).then(|| {
+                ScriptKind::PubKey {
+                    data: v.clone().to_vec().try_into().expect("bounds checked"),
+                }
+            })
+        }
+
+        // Pay-to-Multi-Signature (P2MS)
+        [Opcode::PushValue(PushValue::SmallValue(required)), pubkeys @ .., Opcode::PushValue(PushValue::SmallValue(keys)), op::CHECKMULTISIG] =>
+        {
+            match (u8::try_from(required.to_num()), u8::try_from(keys.to_num())) {
+                (Ok(required @ 0..=16), Ok(keys @ 0..=16)) => {
+                    // The remaining opcodes must be `PushData`s that are valid lengths for pubkeys.
+                    let pubkeys = pubkeys
+                        .iter()
+                        .map(|op| match op {
+                            Opcode::PushValue(data) => {
+                                let pubkey_bytes = data.value();
+                                // Equivalent to `CPubKey::GetLen`
+                                let expected_len = match pubkey_bytes.first() {
+                                    Some(2 | 3) => PubKey::COMPRESSED_SIZE,
+                                    Some(4 | 6 | 7) => PubKey::SIZE,
+                                    _ => 0,
+                                };
+                                // Equivalent to `CPubKey::ValidSize`
+                                (!pubkey_bytes.is_empty() && expected_len == pubkey_bytes.len())
+                                    .then(|| pubkey_bytes.try_into().expect("bounds checked"))
+                            }
+                            _ => None,
+                        })
+                        .collect::<Option<Vec<_>>>();
+
+                    pubkeys.and_then(|pubkeys| {
+                        (pubkeys.len() == usize::from(keys) && required <= keys)
+                            .then_some(ScriptKind::MultiSig { required, pubkeys })
+                    })
+                }
+
+                // Non-standard
+                _ => None,
+            }
+        }
+
+        // Non-standard
+        _ => None,
+    }
+}
+
+/// Known kinds of standard scripts.
+pub enum ScriptKind {
+    /// A P2PKH script.
+    PubKeyHash {
+        /// The Hash160 of the public key.
+        hash: [u8; 20],
```

Comment:
Should we have `type Hash160 = [u8; 20]` somewhere? It’s also used in the `pattern` module.

Reply:
Could be a useful refactor, but for a subsequent PR, and it should be `Hash160Output` or similar (as it's not the hash function itself).